atspi-common/src/object_ref.rs

```diff
@@ -21,75 +23,112 @@ pub struct ObjectRef {
+pub struct ObjectRefBorrowed<'a> {
+	#[serde(borrow)]
+	pub name: UniqueName<'a>,
+	#[serde(borrow)]
+	pub path: ObjectPath<'a>,
+}
+
+impl<'a> ObjectRefBorrowed<'a> {
+	/// Create a new `ObjectRefBorrowed`
+	#[must_use]
+	pub fn new(name: UniqueName<'a>, path: ObjectPath<'a>) -> Self {
+		Self { name, path }
+	}
+
+	/// Convert a partially borrowed `ObjectRefBorrowed` into a fully owned `ObjectRef`
+	// A derived clone would clone the owned fields and create new borrows for the borrowed fields.
+	// Whereas sometimes we want to convert the borrowed fields into owned fields.
+	#[must_use]
+	pub fn to_fully_owned(&self) -> ObjectRef {
+		let name = OwnedUniqueName::from(self.name.clone());
+		let path = OwnedObjectPath::from(self.path.clone());
+		ObjectRef { name, path }
+	}
+
+	/// Create a static `ObjectRefBorrowed`, unchecked, from static string values.
+	///
+	/// # Safety
+	/// The caller must ensure that the strings are valid.
+	#[must_use]
+	pub const fn from_static_str_unchecked(sender: &'static str, path: &'static str) -> Self {
+		Self {
+			name: UniqueName::from_static_str_unchecked(sender),
+			path: ObjectPath::from_static_str_unchecked(path),
+		}
+	}
+}
+
+impl Default for ObjectRefBorrowed<'_> {
+	fn default() -> Self {
+		ObjectRefBorrowed {
+			name: UniqueName::from_static_str_unchecked(":0.0"),
+			path: ObjectPath::from_static_str_unchecked("/org/a11y/atspi/accessible/null"),
+		}
+	}
 }
 
 impl<'a> TryFrom<zvariant::Value<'a>> for ObjectRef {
 	type Error = zvariant::Error;
 	fn try_from(value: zvariant::Value<'a>) -> Result<Self, Self::Error> {
-		value.try_to_owned()?.try_into()
+		// Relies on `TryFrom<OwnedValue> for (T0, T1)` implementation
+		let (name, path): (OwnedUniqueName, OwnedObjectPath) = value.try_to_owned()?.try_into()?;
+		Ok(ObjectRef { name, path })
 	}
 }
 
 impl TryFrom<zvariant::OwnedValue> for ObjectRef {
 	type Error = zvariant::Error;
-	fn try_from<'a>(value: zvariant::OwnedValue) -> Result<Self, Self::Error> {
-		match &*value {
-			zvariant::Value::Structure(s) => {
-				if s.signature() != ObjectRef::SIGNATURE {
-					return Err(zvariant::Error::SignatureMismatch(s.signature().clone(), format!("To turn a zvariant::Value into an atspi::ObjectRef, it must be of type {:?}", ObjectRef::SIGNATURE)));
-				}
-				let fields = s.fields();
-				let name: String =
-					fields.first().ok_or(zvariant::Error::IncorrectType)?.try_into()?;
-				let path_value: ObjectPath<'_> =
-					fields.last().ok_or(zvariant::Error::IncorrectType)?.try_into()?;
-				Ok(ObjectRef {
-					name: name.try_into().map_err(|_| zvariant::Error::IncorrectType)?,
-					path: path_value.into(),
-				})
-			}
-			_ => Err(zvariant::Error::IncorrectType),
-		}
+	fn try_from(value: zvariant::OwnedValue) -> Result<Self, Self::Error> {
+		let (name, path): (OwnedUniqueName, OwnedObjectPath) = value.try_into()?;
+		Ok(ObjectRef { name, path })
+	}
+}
+
+impl<'b, 'a: 'b> TryFrom<Value<'a>> for ObjectRefBorrowed<'b> {
```

Comment:
Do we need two explicit lifetimes? Can we use a single lifetime or elide them instead?

Reply:
You are right. We can do with one.
I tried hard to find an example where the supposed flexibility of two lifetimes was beneficial but could not find any and I currently suspect this is because we are not tied to `Value<'a>` but tied to whatever it was pointing to. Lifetimes are hard, and I am easily misguided still.